File: sdk-typescript/src/store/slotStore.ts
/**
 * The A/B slot store (D35), modelled on OS update partitions.
 *
 *   <stateDir>/airprompter/<agentId>/<target>/
 *     store.json            active slot, staged slot, generation, root document, instanceId, wrapped DEK
 *     slots/A/manifest.json signed envelope, plaintext (no prompt text in it)
 *     slots/A/payloads/sha256-<hex>.enc
 *     slots/B/…
 *
 * Stage into the inactive slot and fsync every file; flip `active` by
 * writing store.json to a temp file and renaming it. A crash mid-apply
 * leaves the previous slot intact and the staged slot either complete or
 * discarded on the next open. Every open verifies the manifest signature
 * against the stored root document and every payload's hash after decrypt.
 * Any failure marks the slot corrupt and the caller falls back: other slot
 * → vendored bundle → refuse to start. Render never serves unverified bytes.
 */

import { randomBytes } from "node:crypto";
import { closeSync, existsSync, fsyncSync, mkdirSync, openSync, readFileSync, readdirSync, renameSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { sha256Prefixed } from "../protocol/canonicalJson.js";
import { referencedPayloads, verifyManifest, type Verdict } from "../protocol/trust.js";
import type { Manifest, RefusalCode, RootMetadata, Target } from "../protocol/types.js";
import type { KeyProvider, StorageProtection } from "./keyProvider.js";
import { decryptPayload, encryptPayload, payloadAad, PayloadDecryptError } from "./payloadCrypto.js";

export type SlotName = "A" | "B";

export interface StoreFile {
  version: 1;
  agentId: string;
  target: Target;
  instanceId: string;
  /** Base64url of the DEK wrapped by the key provider. */
  wrappedDek: string;
  storageProtection: StorageProtection;
  /** The slot that serves; null until the first apply. */
  active: SlotName | null;
  /** A verified, not yet activated release (unlock_required). */
  staged: SlotName | null;
  /** The generation the active slot holds; 0 before the first apply. Anti-rollback compares against it. */
  generation: number;
  /** The last accepted root document; null until one is accepted. */
  root: RootMetadata | null;
  /** Set by a forced local downgrade; reported on evidence. */
  forcedDowngrade?: boolean;
  updatedAt: string;
}

export interface LoadedSlot {
  slot: SlotName;
  manifest: Manifest;
  generation: number;
  signingKeyId: string;
  /** contentHash → plaintext bytes, all verified. */
  payloads: Map<string, Buffer>;
}

export class StoreError extends Error {
  constructor(
    readonly code: "kek_unavailable" | "store_corrupt" | "slot_corrupt" | "generation_rollback" | "no_release" | "not_staged",
    message: string,
    readonly detail?: RefusalCode | string,
  ) {
    super(message);
    this.name = "StoreError";
  }
}

const otherSlot = (slot: SlotName): SlotName => (slot === "A" ? "B" : "A");

function fsyncFile(path: string): void {
  // Write access: on Windows an fsync on a read-only descriptor is refused (FlushFileBuffers needs it).
  const fd = openSync(path, "r+");
  try {
    fsyncSync(fd);
  } finally {
    closeSync(fd);
  }
}

function writeFileSynced(path: string, bytes: Uint8Array, mode = 0o600): void {
  writeFileSync(path, bytes, { mode });
  fsyncFile(path);
}

/** Write to a temp file, fsync, rename: the file is either the old one or the new one, never half. */
function replaceFileAtomically(path: string, bytes: Uint8Array, hooks?: StoreHooks): void {
  const temp = `${path}.${randomBytes(4).toString("hex")}.tmp`;
  writeFileSynced(temp, bytes);
  hooks?.beforeRename?.(path);
  renameSync(temp, path);
  try {
    const dir = openSync(join(path, ".."), "r");
    try {
      fsyncSync(dir);
    } finally {
      closeSync(dir);
    }
  } catch {
    // Directory fsync is best effort on platforms that refuse it (Windows refuses to open a directory this way).
  }
}

/** Test seams: a crash between fsync and rename is the case the layout exists for. */
export interface StoreHooks {
  beforeRename?: (path: string) => void;
}

export interface OpenStoreInput {
  stateDir: string;
  agentId: string;
  target: Target;
  keyProvider: KeyProvider;
  hooks?: StoreHooks;
}

export class SlotStore {
  private constructor(
    readonly dir: string,
    private file: StoreFile,
    private readonly dek: Uint8Array,
    private readonly hooks: StoreHooks | undefined,
  ) {}

  static path(input: { stateDir: string; agentId: string; target: Target }): string {
    return join(input.stateDir, "airprompter", input.agentId, input.target);
  }

  /** Opens (creating on first use). Throws `kek_unavailable` when the key provider cannot produce the KEK. */
  static async open(input: OpenStoreInput): Promise<SlotStore> {
    const dir = SlotStore.path(input);
    mkdirSync(join(dir, "slots", "A", "payloads"), { recursive: true, mode: 0o700 });
    mkdirSync(join(dir, "slots", "B", "payloads"), { recursive: true, mode: 0o700 });
    const storePath = join(dir, "store.json");
    if (!existsSync(storePath)) {
      const dek = randomBytes(32);
      let wrapped: Uint8Array;
      try {
        wrapped = await input.keyProvider.wrap(dek);
      } catch (error) {
        throw new StoreError("kek_unavailable", `the key provider could not wrap the store key: ${(error as Error).message}`);
      }
      const file: StoreFile = {
        version: 1,
        agentId: input.agentId,
        target: input.target,
        instanceId: `i-${randomBytes(12).toString("base64url")}`,
        wrappedDek: Buffer.from(wrapped).toString("base64url"),
        storageProtection: input.keyProvider.storageProtection,
        active: null,
        staged: null,
        generation: 0,
        root: null,
        updatedAt: new Date().toISOString(),
      };
      replaceFileAtomically(storePath, Buffer.from(JSON.stringify(file, null, 2), "utf8"));
      return new SlotStore(dir, file, dek, input.hooks);
    }
    let file: StoreFile;
    try {
      file = JSON.parse(readFileSync(storePath, "utf8")) as StoreFile;
    } catch {
      throw new StoreError("store_corrupt", "store.json is unreadable");
    }
    if (file.version !== 1 || file.agentId !== input.agentId || file.target !== input.target) {
      throw new StoreError("store_corrupt", "store.json belongs to another agent or target");
    }
    let dek: Uint8Array;
    try {
      dek = await input.keyProvider.unwrap(Buffer.from(file.wrappedDek, "base64url"));
    } catch (error) {
      throw new StoreError("kek_unavailable", `the key provider could not unwrap the store key: ${(error as Error).message}`);
    }
    if (dek.length !== 32) throw new StoreError("store_corrupt", "unwrapped store key has the wrong length");
    return new SlotStore(dir, file, dek, input.hooks);
  }

  get state(): Readonly<StoreFile> {
    return this.file;
  }

  get instanceId(): string {
    return this.file.instanceId;
  }

  get storageProtection(): StorageProtection {
    return this.file.storageProtection;
  }

  /**
   * KEK rotation: the DEK is re-WRAPPED under the new provider, never
   * re-generated — every payload stays readable during a live serve and
   * nothing is re-encrypted. The old provider is not consulted (the DEK is
   * already in memory), so a rotation works even after the old KEK is gone.
   */
  async rotateKey(provider: KeyProvider): Promise<void> {
    const wrapped = await provider.wrap(this.dek);
    this.write({ ...this.file, wrappedDek: Buffer.from(wrapped).toString("base64url"), storageProtection: provider.storageProtection });
  }

  /** Persist a newly accepted root document (the caller verified it). */
  acceptRoot(root: RootMetadata): void {
    this.write({ ...this.file, root });
  }

  /**
   * Stage a verified release into the inactive slot: encrypt every payload
   * under the DEK with the slot-binding AAD, write the manifest, fsync all,
   * then record `staged` in store.json. Refuses a generation below the
   * active one unless `force` (a forced downgrade is stamped on evidence).
   */
  stage(input: { manifest: Manifest; payloads: ReadonlyMap<string, Uint8Array>; force?: boolean }): SlotName {
    const generation = input.manifest.payload.generation;
    if (generation < this.file.generation && !input.force) {
      throw new StoreError("generation_rollback", `generation ${generation} is below the stored ${this.file.generation}`);
    }
    for (const [hash, byteLength] of referencedPayloads(input.manifest.payload)) {
      const bytes = input.payloads.get(hash);
      if (!bytes) throw new StoreError("slot_corrupt", `payload ${hash} missing from the stage set`, "payload_missing");
      if (bytes.length !== byteLength || sha256Prefixed(bytes) !== hash) throw new StoreError("slot_corrupt", `payload ${hash} does not hash`, "payload_hash_mismatch");
    }
    const slot = this.file.active ? otherSlot(this.file.active) : "A";
    const slotDir = join(this.dir, "slots", slot);
    rmSync(join(slotDir, "payloads"), { recursive: true, force: true });
    mkdirSync(join(slotDir, "payloads"), { recursive: true, mode: 0o700 });
    for (const [hash, bytes] of input.payloads) {
      const aad = payloadAad({ agentId: this.file.agentId, target: this.file.target, generation, contentHash: hash });
      writeFileSynced(join(slotDir, "payloads", `${hash.replace(":", "-")}.enc`), encryptPayload(this.dek, bytes, aad));
    }
    writeFileSynced(join(slotDir, "manifest.json"), Buffer.from(JSON.stringify(input.manifest), "utf8"));
    this.write({ ...this.file, staged: slot, ...(input.force && generation < this.file.generation ? { forcedDowngrade: true } : {}) });
    return slot;
  }

  /** Flip `active` to the staged slot: one atomic store.json replace. */
  activate(): SlotName {
    const slot = this.file.staged;
    if (!slot) throw new StoreError("not_staged", "nothing is staged");
    const manifest = this.readManifest(slot);
    this.write({ ...this.file, active: slot, staged: null, generation: manifest.payload.generation });
    return slot;
  }

  /** Instant local rollback: the previous release is the other slot. Stamped on evidence by the caller. */
  rollbackLocal(): SlotName {
    if (!this.file.active) throw new StoreError("no_release", "nothing is active");
    const previous = otherSlot(this.file.active);
    const manifest = this.readManifest(previous);
    this.write({ ...this.file, active: previous, staged: null, generation: manifest.payload.generation, forcedDowngrade: manifest.payload.generation < this.file.generation });
    return previous;
  }

  /** Discard a staged slot (a crashed apply, or a refused unlock). */
  discardStaged(): void {
    if (!this.file.staged) return;
    rmSync(join(this.dir, "slots", this.file.staged, "payloads"), { recursive: true, force: true });
    rmSync(join(this.dir, "slots", this.file.staged, "manifest.json"), { force: true });
    this.write({ ...this.file, staged: null });
  }

  /**
   * Load and verify a slot: manifest signature against the stored root,
   * scope, every payload decrypts under this slot's AAD and hashes. Throws
   * `slot_corrupt` on any failure — never returns unverified bytes.
   */
  load(slot: SlotName, input: { now: string; root?: RootMetadata | null; requireCountersign?: boolean; countersignRoot?: RootMetadata | null; expectGeneration?: number }): LoadedSlot {
    const root = input.root ?? this.file.root;
    if (!root) throw new StoreError("slot_corrupt", "no trusted root document to verify against");
    const manifest = this.readManifest(slot);
    const generation = manifest.payload.generation;
    // The active slot must hold the generation store.json says it holds: directories swapped on disk
    // put an older manifest (self-consistent, still signed) behind the active letter, and this is
    // where that shows — the counter lives outside the slots.
    if (input.expectGeneration !== undefined && generation !== input.expectGeneration) {
      throw new StoreError("slot_corrupt", `slot ${slot} holds generation ${generation}, store.json expects ${input.expectGeneration}`, "generation_rollback");
    }
    const payloads = new Map<string, Buffer>();
    for (const hash of referencedPayloads(manifest.payload).keys()) {
      const path = join(this.dir, "slots", slot, "payloads", `${hash.replace(":", "-")}.enc`);
      if (!existsSync(path)) throw new StoreError("slot_corrupt", `payload ${hash} missing`, "payload_missing");
      try {
        payloads.set(hash, decryptPayload(this.dek, readFileSync(path), payloadAad({ agentId: this.file.agentId, target: this.file.target, generation, contentHash: hash })));
      } catch (error) {
        if (error instanceof PayloadDecryptError) throw new StoreError("slot_corrupt", `payload ${hash} does not decrypt for this slot`, "payload_hash_mismatch");
        throw error;
      }
    }
    const verdict: Verdict<{ signingKeyId: string; generation: number }> = verifyManifest({
      manifest,
      root,
      now: input.now,
      scope: { organizationId: manifest.payload.organizationId, agentId: this.file.agentId, target: this.file.target },
      // A stored slot is checked against itself, not the store's counter: the counter is the anti-rollback for NEW manifests.
      storedGeneration: 0,
      payloads,
      countersignRoot: input.countersignRoot ?? null,
      ...(input.requireCountersign !== undefined ? { requireCountersign: input.requireCountersign } : {}),
    });
    if (!verdict.ok) {
      // Expired root: the bytes are still the last verified release; report and serve (D39). Everything else is corrupt.
      if (verdict.reason === "root_expired") {
        const relaxed = verifyManifest({ manifest, root: { ...root, signed: { ...root.signed, expires: "9999-12-31T23:59:59Z" } }, now: input.now, scope: { organizationId: manifest.payload.organizationId, agentId: this.file.agentId, target: this.file.target }, storedGeneration: 0, payloads });
        if (relaxed.ok) return { slot, manifest, generation, signingKeyId: relaxed.signingKeyId, payloads };
      }
      throw new StoreError("slot_corrupt", `slot ${slot} failed verification: ${verdict.reason}`, verdict.reason);
    }
    return { slot, manifest, generation, signingKeyId: verdict.signingKeyId, payloads };
  }

  listSlotFiles(slot: SlotName): string[] {
    const dir = join(this.dir, "slots", slot);
    return existsSync(dir) ? readdirSync(dir, { recursive: true }).map(String).sort() : [];
  }

  private readManifest(slot: SlotName): Manifest {
    const path = join(this.dir, "slots", slot, "manifest.json");
    if (!existsSync(path)) throw new StoreError("slot_corrupt", `slot ${slot} has no manifest`);
    try {
      return JSON.parse(readFileSync(path, "utf8")) as Manifest;
    } catch {
      throw new StoreError("slot_corrupt", `slot ${slot} manifest is unreadable`, "schema_invalid");
    }
  }

  private write(next: StoreFile): void {
    const file = { ...next, updatedAt: new Date().toISOString() };
    replaceFileAtomically(join(this.dir, "store.json"), Buffer.from(JSON.stringify(file, null, 2), "utf8"), this.hooks);
    this.file = file;
  }
}

File: sdk-python/src/airprompter_agent/store/slot_store.py
"""The A/B slot store (D35), modelled on OS update partitions.

::

    <state_dir>/airprompter/<agentId>/<target>/
      store.json            active slot, staged slot, generation, root document, instanceId, wrapped DEK
      slots/A/manifest.json signed envelope, plaintext (no prompt text in it)
      slots/A/payloads/sha256-<hex>.enc
      slots/B/…

Stage into the inactive slot and fsync every file; flip ``active`` by
writing store.json to a temp file and renaming it. A crash mid-apply
leaves the previous slot intact and the staged slot either complete or
discarded on the next open. Every open verifies the manifest signature
against the stored root document and every payload's hash after decrypt.
Any failure marks the slot corrupt and the caller falls back: other slot
→ vendored bundle → refuse to start. Render never serves unverified bytes.

The on-disk layout is byte-for-byte the TypeScript SDK's, so a store one
SDK wrote is a store the other (and the host daemon) can open.
"""

from __future__ import annotations

import json
import os
import shutil
from dataclasses import dataclass, field
from typing import Any, Callable, Mapping, Optional

from .._util import b64url_decode, b64url_encode, fsync_dir, iso_ms, now_ms, random_id
from ..protocol.canonical_json import sha256_prefixed
from ..protocol.trust import Verdict, referenced_payloads, verify_manifest
from .key_provider import KeyProvider
from .payload_crypto import PayloadDecryptError, decrypt_payload, encrypt_payload, payload_aad

SlotName = str  # "A" | "B"


@dataclass
class LoadedSlot:
    slot: SlotName
    manifest: dict[str, Any]
    generation: int
    signing_key_id: str
    #: contentHash → plaintext bytes, all verified.
    payloads: dict[str, bytes] = field(default_factory=dict)


class StoreError(Exception):
    def __init__(self, code: str, message: str, detail: Optional[str] = None):
        # code: "kek_unavailable" | "store_corrupt" | "store_newer" | "slot_corrupt" | "generation_rollback" | "no_release" | "not_staged"
        super().__init__(message)
        self.code = code
        self.detail = detail


def _other_slot(slot: SlotName) -> SlotName:
    return "B" if slot == "A" else "A"


def _write_file_synced(path: str, data: bytes, mode: int = 0o600) -> None:
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, mode)
    try:
        os.write(fd, data)
        os.fsync(fd)
    finally:
        os.close(fd)


def _replace_file_atomically(path: str, data: bytes, before_rename: Optional[Callable[[str], None]] = None) -> None:
    """Write to a temp file, fsync, rename: the file is either the old one or the new one, never half."""
    temp = f"{path}.{os.urandom(4).hex()}.tmp"
    _write_file_synced(temp, data)
    if before_rename:
        before_rename(path)
    os.replace(temp, path)
    fsync_dir(os.path.dirname(path))


@dataclass
class StoreHooks:
    """Test seams: a crash between fsync and rename is the case the layout exists for."""

    before_rename: Optional[Callable[[str], None]] = None
    #: S8: the writer this runtime records in store.json (the SDK's or the daemon's name and version).
    writer: Optional[Mapping[str, str]] = None


#: S8: store.json is a cross-package contract (``protocol/store-format.md``): a reader at format N accepts N and N-1, writes N,
#: migrates an N-1 file forward on its first write, and refuses N+1 with ``store_newer`` naming the writer.
STORE_FORMAT_VERSION = 2
STORE_FORMATS_READ = frozenset({1, 2})
DEFAULT_WRITER = {"name": "agent-sdk-python", "version": "0.1.0"}


class SlotStore:
    def __init__(self, directory: str, file: dict[str, Any], dek: bytes, hooks: Optional[StoreHooks]):
        self.dir = directory
        self._file = file
        self._dek = dek
        self._hooks = hooks

    @staticmethod
    def path(*, state_dir: str, agent_id: str, target: str) -> str:
        return os.path.join(state_dir, "airprompter", agent_id, target)

    @classmethod
    def open(cls, *, state_dir: str, agent_id: str, target: str, key_provider: KeyProvider, hooks: Optional[StoreHooks] = None) -> "SlotStore":
        """Opens (creating on first use). Raises ``kek_unavailable`` when the key provider cannot produce the KEK."""
        directory = cls.path(state_dir=state_dir, agent_id=agent_id, target=target)
        for slot in ("A", "B"):
            os.makedirs(os.path.join(directory, "slots", slot, "payloads"), mode=0o700, exist_ok=True)
        store_path = os.path.join(directory, "store.json")
        if not os.path.exists(store_path):
            dek = os.urandom(32)
            try:
                wrapped = key_provider.wrap(dek)
            except Exception as error:  # noqa: BLE001 — whatever the provider raised is the reason
                raise StoreError("kek_unavailable", f"the key provider could not wrap the store key: {error}") from error
            file: dict[str, Any] = {
                "version": STORE_FORMAT_VERSION,
                "writer": dict(hooks.writer) if hooks and hooks.writer else dict(DEFAULT_WRITER),
                "agentId": agent_id,
                "target": target,
                "instanceId": random_id(),
                "wrappedDek": b64url_encode(wrapped),
                "storageProtection": key_provider.storage_protection,
                "active": None,
                "staged": None,
                "generation": 0,
                "root": None,
                "updatedAt": iso_ms(now_ms()),
            }
            _replace_file_atomically(store_path, json.dumps(file, indent=2).encode("utf-8"))
            return cls(directory, file, dek, hooks)
        try:
            with open(store_path, encoding="utf-8") as f:
                file = json.load(f)
        except (OSError, ValueError) as error:
            raise StoreError("store_corrupt", "store.json is unreadable") from error
        version = file.get("version")
        if isinstance(version, int) and not isinstance(version, bool) and version > STORE_FORMAT_VERSION:
            # N+1: written by something newer than this reader. Never guessed at; the writer is named so the operator knows what to update.
            writer = file.get("writer")
            named = f"{writer.get('name')} {writer.get('version')}" if isinstance(writer, Mapping) else "an unknown writer"
            raise StoreError("store_newer", f"store.json is format {version}, written by {named}; this runtime reads formats {' and '.join(str(v) for v in sorted(STORE_FORMATS_READ))} — update it, or roll the writer back before its next write", named)
        if version not in STORE_FORMATS_READ or file.get("agentId") != agent_id or file.get("target") != target:
            raise StoreError("store_corrupt", "store.json belongs to another agent or target")
        try:
            dek = key_provider.unwrap(b64url_decode(file["wrappedDek"]))
        except Exception as error:  # noqa: BLE001
            raise StoreError("kek_unavailable", f"the key provider could not unwrap the store key: {error}") from error
        if len(dek) != 32:
            raise StoreError("store_corrupt", "unwrapped store key has the wrong length")
        return cls(directory, file, dek, hooks)

    @property
    def state(self) -> Mapping[str, Any]:
        return self._file

    @property
    def instance_id(self) -> str:
        return self._file["instanceId"]

    @property
    def storage_protection(self) -> str:
        return self._file["storageProtection"]

    def rotate_key(self, provider: KeyProvider) -> None:
        """KEK rotation: the DEK is re-WRAPPED under the new provider, never re-generated — every payload stays readable and nothing is re-encrypted."""
        wrapped = provider.wrap(self._dek)
        self._write({**self._file, "wrappedDek": b64url_encode(wrapped), "storageProtection": provider.storage_protection})

    def accept_root(self, root: Mapping[str, Any]) -> None:
        """Persist a newly accepted root document (the caller verified it)."""
        self._write({**self._file, "root": dict(root)})

    def pin_apply_policy(self, *, value: str, source: str, generation: int, set_at: str) -> None:
        """S4: record the apply policy this host holds — ``{"value", "source": "manifest" | "operator", "generation", "setAt"}``."""
        self._write({**self._file, "applyPolicyPin": {"value": value, "source": source, "generation": generation, "setAt": set_at}})

    def stage(self, *, manifest: Mapping[str, Any], payloads: Mapping[str, bytes], force: bool = False) -> SlotName:
        """Stage a verified release into the inactive slot. Refuses a generation below the active one unless ``force`` (a forced downgrade is stamped on evidence)."""
        generation = manifest["payload"]["generation"]
        if generation < self._file["generation"] and not force:
            raise StoreError("generation_rollback", f"generation {generation} is below the stored {self._file['generation']}")
        for content_hash, byte_length in referenced_payloads(manifest["payload"]).items():
            data = payloads.get(content_hash)
            if data is None:
                raise StoreError("slot_corrupt", f"payload {content_hash} missing from the stage set", "payload_missing")
            if len(data) != byte_length or sha256_prefixed(data) != content_hash:
                raise StoreError("slot_corrupt", f"payload {content_hash} does not hash", "payload_hash_mismatch")
        slot = _other_slot(self._file["active"]) if self._file["active"] else "A"
        slot_dir = os.path.join(self.dir, "slots", slot)
        shutil.rmtree(os.path.join(slot_dir, "payloads"), ignore_errors=True)
        os.makedirs(os.path.join(slot_dir, "payloads"), mode=0o700, exist_ok=True)
        for content_hash, data in payloads.items():
            aad = payload_aad(agent_id=self._file["agentId"], target=self._file["target"], generation=generation, content_hash=content_hash)
            _write_file_synced(os.path.join(slot_dir, "payloads", f"{content_hash.replace(':', '-')}.enc"), encrypt_payload(self._dek, data, aad))
        _write_file_synced(os.path.join(slot_dir, "manifest.json"), json.dumps(manifest, separators=(",", ":")).encode("utf-8"))
        next_file = {**self._file, "staged": slot}
        if force and generation < self._file["generation"]:
            next_file["forcedDowngrade"] = True
        self._write(next_file)
        return slot

    def activate(self) -> SlotName:
        """Flip ``active`` to the staged slot: one atomic store.json replace."""
        slot = self._file.get("staged")
        if not slot:
            raise StoreError("not_staged", "nothing is staged")
        generation = self._read_manifest(slot)["payload"]["generation"]
        next_file = dict(self._file)
        held = next_file.get("heldBackBelow")
        if held is not None and generation > held:
            del next_file["heldBackBelow"]  # Moving past a held-back generation ends the hold; a forced downgrade stays stamped.
        next_file.update({"active": slot, "staged": None, "generation": generation})
        self._write(next_file)
        return slot

    def rollback_local(self) -> SlotName:
        """Instant local rollback: the previous release is the other slot. Stamped on evidence by the caller."""
        if not self._file.get("active"):
            raise StoreError("no_release", "nothing is active")
        previous = _other_slot(self._file["active"])
        generation = self._read_manifest(previous)["payload"]["generation"]
        downgrade = generation < self._file["generation"]
        next_file = {**self._file, "active": previous, "staged": None, "generation": generation, "forcedDowngrade": downgrade}
        if downgrade:
            next_file["heldBackBelow"] = self._file["generation"]
        self._write(next_file)
        return previous

    def discard_staged(self) -> None:
        """Discard a staged slot (a crashed apply, or a refused unlock)."""
        slot = self._file.get("staged")
        if not slot:
            return
        shutil.rmtree(os.path.join(self.dir, "slots", slot, "payloads"), ignore_errors=True)
        try:
            os.remove(os.path.join(self.dir, "slots", slot, "manifest.json"))
        except FileNotFoundError:
            pass
        self._write({**self._file, "staged": None})

    def load(
        self,
        slot: SlotName,
        *,
        now: str,
        root: Optional[Mapping[str, Any]] = None,
        require_countersign: Optional[bool] = None,
        countersign_root: Optional[Mapping[str, Any]] = None,
        expect_generation: Optional[int] = None,
    ) -> LoadedSlot:
        """Load and verify a slot. Raises ``slot_corrupt`` on any failure — never returns unverified bytes."""
        trusted = root if root is not None else self._file.get("root")
        if not trusted:
            raise StoreError("slot_corrupt", "no trusted root document to verify against")
        manifest = self._read_manifest(slot)
        generation = manifest["payload"]["generation"]
        # The active slot must hold the generation store.json says it holds: directories swapped on disk put an older
        # manifest (self-consistent, still signed) behind the active letter, and this is where that shows.
        if expect_generation is not None and generation != expect_generation:
            raise StoreError("slot_corrupt", f"slot {slot} holds generation {generation}, store.json expects {expect_generation}", "generation_rollback")
        payloads: dict[str, bytes] = {}
        for content_hash in referenced_payloads(manifest["payload"]):
            path = os.path.join(self.dir, "slots", slot, "payloads", f"{content_hash.replace(':', '-')}.enc")
            if not os.path.exists(path):
                raise StoreError("slot_corrupt", f"payload {content_hash} missing", "payload_missing")
            with open(path, "rb") as f:
                sealed = f.read()
            try:
                payloads[content_hash] = decrypt_payload(self._dek, sealed, payload_aad(agent_id=self._file["agentId"], target=self._file["target"], generation=generation, content_hash=content_hash))
            except PayloadDecryptError as error:
                raise StoreError("slot_corrupt", f"payload {content_hash} does not decrypt for this slot", "payload_hash_mismatch") from error
        scope = {"organizationId": manifest["payload"]["organizationId"], "agentId": self._file["agentId"], "target": self._file["target"]}
        # A stored slot is checked against itself, not the store's counter: the counter is the anti-rollback for NEW manifests.
        verdict: Verdict = verify_manifest(manifest=manifest, root=trusted, now=now, scope=scope, stored_generation=0, payloads=payloads, countersign_root=countersign_root, require_countersign=require_countersign)
        if not verdict.ok:
            # Expired root: the bytes are still the last verified release; report and serve (D39). Everything else is corrupt.
            if verdict.reason == "root_expired":
                relaxed_root = {**trusted, "signed": {**trusted["signed"], "expires": "9999-12-31T23:59:59Z"}}
                relaxed = verify_manifest(manifest=manifest, root=relaxed_root, now=now, scope=scope, stored_generation=0, payloads=payloads)
                if relaxed.ok:
                    return LoadedSlot(slot, manifest, generation, relaxed.signing_key_id or "", payloads)
            raise StoreError("slot_corrupt", f"slot {slot} failed verification: {verdict.reason}", verdict.reason)
        return LoadedSlot(slot, manifest, generation, verdict.signing_key_id or "", payloads)

    def list_slot_files(self, slot: SlotName) -> list[str]:
        base = os.path.join(self.dir, "slots", slot)
        if not os.path.isdir(base):
            return []
        names = []
        for root_dir, _dirs, files in os.walk(base):
            for name in files:
                names.append(os.path.relpath(os.path.join(root_dir, name), base))
        return sorted(names)

    def _read_manifest(self, slot: SlotName) -> dict[str, Any]:
        path = os.path.join(self.dir, "slots", slot, "manifest.json")
        if not os.path.exists(path):
            raise StoreError("slot_corrupt", f"slot {slot} has no manifest")
        try:
            with open(path, encoding="utf-8") as f:
                return json.load(f)
        except (OSError, ValueError) as error:
            raise StoreError("slot_corrupt", f"slot {slot} manifest is unreadable", "schema_invalid") from error

    def _write(self, next_file: dict[str, Any]) -> None:
        # S8: what this reader would write — an N-1 file migrates forward here, on the first write, never on open (the rollback window).
        file = {**next_file, "version": STORE_FORMAT_VERSION, "writer": dict(self._hooks.writer) if self._hooks and self._hooks.writer else dict(DEFAULT_WRITER), "updatedAt": iso_ms(now_ms())}
        _replace_file_atomically(os.path.join(self.dir, "store.json"), json.dumps(file, indent=2).encode("utf-8"), self._hooks.before_rename if self._hooks else None)
        self._file = file
